### PR TITLE
fix weight rerank of knowledge retrieval

### LINKED ISSUE #10930: weight rerank not working properly in knowledge retrieval
### Self Checks

- [X] This is only for bug report, if you would like to ask a question, please head to [Discussions](https://github.com/langgenius/dify/discussions/categories/general).
- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### Dify version

main

### Cloud or Self Hosted

Self Hosted (Source)

### Steps to reproduce

1. configure retrieval method to hybrid and use weight rerank model
2. use hit testing 
3. you will see empty result

### ✔️ Expected Behavior

return some chunks

### ❌ Actual Behavior

empty

### PR DESCRIPTION
**Pull request by AkisAya**
# Summary

fixes #10930 

> [!Tip]
> Close issue syntax: `Fixes #<issue number>` or `Resolves #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

# Screenshots

<table>
  <tr>
  <td>Before: </td>
  <td>After: </td>
  </tr>
  <tr>
  <td>...</td>
  <td>...</td>
  </tr>
</table>

# Checklist

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

### CODE REVIEW

**Comment by crazywoola** [comment]
@yihong0618 Can you take a look as well.

**Review by yihong0618 — APPROVED** [review]
LGTM
thanks for the change.

**Review by bowenliang123 — APPROVED** [review]

File: api/core/rag/rerank/weight_rerank.py [hunk 1]
```diff
@@ -39,9 +39,8 @@ def run(
         unique_documents = []
         doc_id = set()
```
Comment (bowenliang123) [inline api/core/rag/rerank/weight_rerank.py]: How about change the name into `doc_ids` to prevent ambiguous usage?

Comment (AkisAya) [reply api/core/rag/rerank/weight_rerank.py]: sure, doc_ids is better

File: api/core/rag/rerank/weight_rerank.py [hunk 2]
```diff
@@ -37,11 +37,10 @@ def run(
...
         unique_documents = []
-        doc_id = set()
+        doc_ids = set()
         for document in documents:
-            doc_id = document.metadata.get("doc_id")
-            if doc_id not in doc_id:
-                doc_id.add(doc_id)
+            if document.metadata["doc_id"] not in doc_ids:
+                doc_ids.add(document.metadata["doc_id"])
                 unique_documents.append(document)
```
Comment (bowenliang123) [inline api/core/rag/rerank/weight_rerank.py:44]: Here's another optimised version in single-line code, with an equivalent effect for all these lines. ✌🏻

`unique_documents = {doc.metadata["doc_id"]: doc for doc in documents}.values()`

Comment (AkisAya) [reply api/core/rag/rerank/weight_rerank.py:44]: yes, it's pythonic

Comment (AkisAya) [reply api/core/rag/rerank/weight_rerank.py:44]: in rerank_model.py which dealt with document.provider == "external"
```
elif document.provider == "external":
    if document not in unique_documents:
        unique_documents.append(document)
```
is this the only way to judge whether a external document duplicate?
does a external document has a doc_id in metadata?

Comment (bowenliang123) [reply api/core/rag/rerank/weight_rerank.py:44]: Just a piece of advice. We could ignore it.
LGTM for now.

**Comment by AkisAya** [comment]
> @AkisAya @yihong0618 Can you guys add my wechat. We have a wechat groups for contributors. :)

ok，how to get your wechat num

**Comment by crazywoola** [comment]
crazyphage @AkisAya

**Review by crazywoola — APPROVED** [review]

**Comment by AkisAya** [comment]
> crazyphage @AkisAya
applied

**Pull request merged**